### PR TITLE
TIP-1159: Do not store missing required attributes anymore

### PR DESCRIPTION
**Pull request by mmetayer**
The only place where we need to get missing required attributes for a product is in the PEF (meaning internal_api normalizer, for a single product). Storing these missing attributes in DB is very costly, it is more efficient to re-calculate the missing attributes on the fly while normalizing the product.

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | OK
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/Projection/ProductCompleteness.php [hunk 1]
```diff
@@ -20,23 +20,22 @@ final class ProductCompleteness
...
-        array $missingAttributeCodes
-    ) {
+    public function __construct(string $channelCode, string $localeCode, int $requiredCount, int $missingCount)
+    {
         if ($requiredCount < 0) {
             throw new \InvalidArgumentException('$requiredCount expects a positive integer');
         }
+        if ($missingCount < 0) {
+            throw new \InvalidArgumentException('$missingCount expects a positive integer');
+        }
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Model/Projection/ProductCompleteness.php]: You should add a condition $required should be >= $missing (and a test for that)

File: src/Akeneo/Pim/Enrichment/Component/Product/Normalizer/InternalApi/ProductNormalizer.php [hunk 2]
```diff
@@ -262,15 +263,17 @@ function ($group) {
     }
 
     /**
-     * Get Product Completeness and normalize it
+     * Calculates the product completenesses, and normalizes them
+     * It is costly, but needed to get the missing attribute codes as we do not store them anymore
+     * Also, this normlizer is (and should be) only used to normalize a single product for the PEF
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Normalizer/InternalApi/ProductNormalizer.php]: normlizer

File: src/Akeneo/Pim/Enrichment/Component/Product/Normalizer/InternalApi/ProductNormalizer.php [hunk 3]
```diff
@@ -61,16 +59,7 @@ class ProductNormalizer implements NormalizerInterface
     /** @var ConverterInterface */
     protected $productValueConverter;
 
-    /** @var ObjectManager */
-    protected $productManager;
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Normalizer/InternalApi/ProductNormalizer.php]: These dependencies which were not used anymore

File: tests/back/Pim/Enrichment/Specification/Component/Product/Model/Projection/ProductCompletenessWithMissingAttributeCodesCollectionSpec.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    {
+        $completeness = new ProductCompletenessWithMissingAttributeCodes('ecommerce', 'en_US', 4, []);
+        $otherCompleteness = new ProductCompletenessWithMissingAttributeCodes('ecommerce', 'en_US', 5, ['description', 'price']);
+        $this->beConstructedWith(42, [$completeness, $otherCompleteness]);
+
+        $this->getIterator()->count()->shouldReturn(1);
+        $this->getIterator()->getArrayCopy()->shouldReturn(['ecommerce-en_US' => $otherCompleteness]);
+    }
+
+    function it_can_retriev_a_completeness_by_channel_and_locale()
```
Comment (ahocquard) [inline tests/back/Pim/Enrichment/Specification/Component/Product/Model/Projection/ProductCompletenessWithMissingAttributeCodesCollectionSpec.php]:
```suggestion
    function it_can_retrieve_a_completeness_by_channel_and_locale()
```

File: tests/legacy/features/pim/enrichment/product/completeness/remove_completeness.feature [hunk 5]
```diff
@@ -43,15 +43,13 @@ Feature: Display the completeness of a product
...
       | channel | locale | state   | missing_values | ratio |
-      | tablet  | en_US  | warning | 1              | 88%   |
       | tablet  | fr_FR  | warning | 2              | 77%   |
       | mobile  | en_US  | success | 0              | 100%  |
       | mobile  | fr_FR  | success | 0              | 100%  |
     When I am on the "sandals" product page
     And I visit the "Completeness" column tab
     Then I should see the completeness:
       | channel | locale | state   | missing_values | ratio |
-      | tablet  | en_US  | warning | 6              | 25%   |
```
Comment (ahocquard) [inline tests/legacy/features/pim/enrichment/product/completeness/remove_completeness.feature]: was it a bug?

Comment (mmetayer) [reply tests/legacy/features/pim/enrichment/product/completeness/remove_completeness.feature]: It used to be a bug, it's not anymore :)

**Pull request merged**